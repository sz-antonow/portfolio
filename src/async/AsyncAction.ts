import { ActionMeta } from 'redux-actions';



export type AsyncAction<Payload> = ActionMeta<Payload, AsyncActionMeta>;

export namespace AsyncAction {
  /**
   * Creates action type
   * from given action type and current async status
   * @param type
   * @param asyncStatus
   */
  export const createType = (type: string, asyncStatus: AsyncStatus): string => `${type}_${asyncStatus}`;

  export type CreateTypes = {
    VOID: string;
    PENDING: string;
    RESOLVED: string;
    REJECTED: string;
  };
  export const createTypes = (type: string): CreateTypes => ({
    VOID: createType(type, AsyncStatus.VOID),
    PENDING: createType(type, AsyncStatus.PENDING),
    RESOLVED: createType(type, AsyncStatus.RESOLVED),
    REJECTED: createType(type, AsyncStatus.REJECTED)
  });

  /**
   * Action init for async actions init
   * returns action marked as async
   * @param type
   * @param asyncStatus
   * @param payload
   */
  export function create<Payload>(type: string, asyncStatus: AsyncStatus, payload: Payload): AsyncAction<Payload> {
    const action = {
      type: createType(type, asyncStatus),
      payload,
      meta: {
        type,
        asyncStatus: asyncStatus
      }
    };

    return asyncStatus === AsyncStatus.REJECTED ? { ...action, error: true } : action;
  }
}
