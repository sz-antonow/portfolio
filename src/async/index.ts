export { AsyncAction } from './AsyncAction';
import { AsyncActionMeta as AsAcM } from './AsyncActionMeta';
export { AsyncActionCreator } from './AsyncActionCreator';
export { AsyncState } from './AsyncState';
export { AsyncSelector } from './AsyncSelector';

export type AsyncActionMeta = AsAcM;
