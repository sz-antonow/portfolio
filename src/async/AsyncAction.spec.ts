import { AsyncStatus } from 'common/src/enum';
import { AsyncAction } from './AsyncAction';

describe('AsyncAction', () => {
  it('should create type with current async status', () => {
    const type = 'TYPE';

    expect(AsyncAction.createType(type, AsyncStatus.PENDING)).toMatchSnapshot();
    expect(AsyncAction.createType(type, AsyncStatus.RESOLVED)).toMatchSnapshot();
    expect(AsyncAction.createType(type, AsyncStatus.REJECTED)).toMatchSnapshot();
  });

  it('should create action marked as async', () => {
    const type = 'TYPE';
    const asyncType = AsyncStatus.PENDING;
    const payload = ['foo'];
    const action = AsyncAction.create(type, asyncType, payload);

    expect(action.meta).toHaveProperty('type');
    expect(action.meta).toHaveProperty('asyncStatus');
  });

  it('should create resolved action with given payload', () => {
    const type = 'TYPE';
    const asyncType = AsyncStatus.RESOLVED;
    const payload = ['foo'];
    const action = AsyncAction.create(type, asyncType, payload);

    expect(action.payload).toEqual(payload);
  });

  it('should create pending action', () => {
    const type = 'TYPE';
    const asyncType = AsyncStatus.PENDING;
    const action = AsyncAction.create(type, asyncType, null);

    expect(action).toMatchSnapshot();
  });

  it('should create resolved action', () => {
    const type = 'TYPE';
    const asyncType = AsyncStatus.RESOLVED;
    const payload = ['foo'];
    const action = AsyncAction.create(type, asyncType, payload);

    expect(action).toMatchSnapshot();
  });

  it('should create rejected action', () => {
    const type = 'TYPE';
    const asyncType = AsyncStatus.REJECTED;
    const payload = new Error('Error message');
    const action = AsyncAction.create(type, asyncType, payload);

    expect(action).toMatchSnapshot();
  });
});
