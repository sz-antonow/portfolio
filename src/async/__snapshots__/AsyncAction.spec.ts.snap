// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AsyncAction should create pending action 1`] = `
Object {
  "meta": Object {
    "asyncStatus": "PENDING",
    "type": "TYPE",
  },
  "payload": null,
  "type": "TYPE_PENDING",
}
`;

exports[`AsyncAction should create rejected action 1`] = `
Object {
  "error": true,
  "meta": Object {
    "asyncStatus": "REJECTED",
    "type": "TYPE",
  },
  "payload": [Error: Error message],
  "type": "TYPE_REJECTED",
}
`;

exports[`AsyncAction should create resolved action 1`] = `
Object {
  "meta": Object {
    "asyncStatus": "RESOLVED",
    "type": "TYPE",
  },
  "payload": Array [
    "foo",
  ],
  "type": "TYPE_RESOLVED",
}
`;

exports[`AsyncAction should create type with current async status 1`] = `"TYPE_PENDING"`;

exports[`AsyncAction should create type with current async status 2`] = `"TYPE_RESOLVED"`;

exports[`AsyncAction should create type with current async status 3`] = `"TYPE_REJECTED"`;
